src/services/refactors/extractType.ts

```diff
@@ -0,0 +1,141 @@
+            return {
+                isJS,
+                selection,
+                firstStatement: Debug.assertDefined((findAncestor(selection, isStatementButNotBlockAndHasJSDoc))),
+            };
+        }
+        else {
+            const firstStatement = Debug.assertDefined((findAncestor(selection, isStatementButNotBlock)));
+            const typeParameters = checkAndCollectionTypeArguments(context.program.getTypeChecker(), selection, firstStatement, file);
+            if (!typeParameters) return undefined;
+
+            return { isJS, selection, firstStatement, typeParameters };
+        }
+    }
+
+    function isStatementButNotBlock(n: Node): n is Statement {
+        return n && isStatement(n) && !isBlock(n);
+    }
+
+    function isStatementButNotBlockAndHasJSDoc(n: Node): n is (Statement & HasJSDoc) {
+        return isStatementButNotBlock && hasJSDocNodes(n);
+    }
+
+    function rangeContainsSkipTrivia(r1: TextRange, node: Node, file: SourceFile): boolean {
+        return rangeContainsStartEnd(r1, skipTrivia(file.text, node.pos), node.end);
+    }
+
+    function checkAndCollectionTypeArguments(checker: TypeChecker, selection: TypeNode, statement: Statement, file: SourceFile): string[] | undefined {
+        let hasError = false;
+        const result: string[] = [];
+        visitor(selection);
+        return hasError ? undefined : result;
+
+        function visitor(node: Node) {
+            if (isTypeReferenceNode(node)) {
+                if (isIdentifier(node.typeName)) {
+                    const symbol = checker.resolveName(node.typeName.text, node.typeName, SymbolFlags.TypeParameter, /* excludeGlobals */ true);
+                    if (symbol) {
+                        const declaration = first(symbol.declarations);
+                        if (rangeContainsSkipTrivia(statement, declaration, file) && !rangeContainsSkipTrivia(selection, declaration, file)) {
+                            result.push(node.typeName.text);
+                        }
+                    }
+                }
+            }
+            else if (isInferTypeNode(node)) {
+                const conditionalTypeNode = findAncestor(node, isConditionalTypeNode);
+                if (!conditionalTypeNode || !rangeContainsSkipTrivia(selection, conditionalTypeNode, file)) {
+                    hasError = true;
+                    return;
+                }
+            }
+            else if (isTypePredicateNode(node) && !rangeContainsSkipTrivia(selection, node.parent, file)) {
+                hasError = true;
+                return;
+            }
+            else if (isTypeQueryNode(node) && isIdentifier(node.exprName)) {
+                const symbol = checker.resolveName(node.exprName.text, node.exprName, SymbolFlags.Value, /* excludeGlobals */ false);
+                if (symbol && rangeContainsSkipTrivia(statement, symbol.valueDeclaration, file) && !rangeContainsSkipTrivia(selection, symbol.valueDeclaration, file)) {

```

Comment:
why are you using `symbol.valueDeclaration` here but using `first(symbol.declarations)` above?

Reply:
If i'm right, using `first(symbol.declarations)` above caused that is a `TypeParameter`, that do not have a `value`  
Using `symbol.valueDeclaration` causes this is a `value`
(the `SymbolFlags.Value` in the `resolveName` call)

Reply:
that makes sense, thanks